File: Swift/AoC 2023/AoC 2023/Solutions/Day12.swift
import Foundation

class Day12: AoCSolution {
    override init() {
        super.init()
        self.day = 12
        self.name = "Hot Springs"
        self.emptyLinesIndicateMultipleInputs = true
    }
    
    override func solve(_ input: AoCInput) -> AoCResult {
        super.solve(input)
        
        let records = input.textLines.map {SpringRecord(s: $0)}
        var part1 = 0
		for i in 0..<records.count {
			var record = records[i]
			part1 += record.possibleArrangementCount()
		}
		
		let expandedRecords = records.map { $0.expanded(times: 5) }
//		print(expandedRecords)
		var part2 = 0
		for i in 0..<expandedRecords.count {
			var record = expandedRecords[i]
			part2 += record.possibleArrangementCount()
		}

        return AoCResult(part1: "\(part1)", part2: "\(part2)")
    }
}

enum SpringCondition {
    case operational
    case damaged
    case unknown
	
	static func from(stringRepresentation s:String) -> SpringCondition {
		switch s {
		case ".":
			return .operational
		case "#":
			return .damaged
		default:
			return .unknown
		}
	}
	
	var stringRepresentation: String {
		switch self {
		case .damaged: return "#"
		case .operational: return "."
		default: return "?"
		}
	}
}

struct MemoRec: Hashable {
	let p: Int
	let g: Int
}

struct SpringRecord: CustomDebugStringConvertible {
    let conditionRecord: [SpringCondition]
    let groupLengths: [Int]
	private let originalConditionRecord: String
	private let originalGroupLengths: String

    init(s: String) {
        let parts = s.split(separator: " ")
		var cr = parts[0].map({SpringCondition.from(stringRepresentation: String($0))})
		cr.append(.operational) // Add operational spring at the end
		conditionRecord = cr
        groupLengths = parts[1].split(separator: ",").map {Int(String($0))!}
		originalConditionRecord = String(parts[0])
		originalGroupLengths = String(parts[1])
    }
    
	private var _memo = Dictionary<MemoRec,Int>()
    mutating func possibleArrangementCount() -> Int {
		_memo.removeAll()
		return arrangementCount(p: 0, g: 0)
    }
	
	//https://github.com/vanam/CodeUnveiled/blob/master/Advent%20Of%20Code%202023/12/main.py
	mutating func arrangementCount(p: Int, g: Int) -> Int {
		if g >= groupLengths.count { // no more groups
			if p < conditionRecord.count && conditionRecord[p...].contains(.damaged) {
				// eg: .##?????#.. 4,1
				return 0 // # not a solution - there are still damaged springs in the record
			}
			return 1
		}
		
		let gs = groupLengths[g] // damaged group size
		if p+gs >= conditionRecord.count { // This used to be just p >=, but line 109 could be out of range
			return 0 // we ran out of springs but there are still groups to arrange
		}

		let mRec = MemoRec(p: p, g: g)
		if _memo.keys.contains(mRec) { return _memo[mRec]! }

		var result = 0

		if conditionRecord[p] == .unknown {
			// if we can start group of damaged springs here
			// eg: '??#...... 3' we can place 3 '#' and there is '?' or '.' after the group
			// eg: '??##...... 3' we cannot place 3 '#' here
			if !conditionRecord[p..<p+gs].contains(.operational) &&
				conditionRecord[p + gs] != .damaged {
				// start damaged group here + this spring is operational ('.')
				result = arrangementCount(p: p + gs + 1, g: g + 1) + arrangementCount(p: p + 1, g: g)
			}
			else {
				// this spring is operational ('.')
				result = arrangementCount(p: p + 1, g: g)
			}
		}
		else if conditionRecord[p] == .damaged {
			// if we can start damaged group here
			if !conditionRecord[p..<p + gs].contains(.operational) && conditionRecord[p + gs] != .damaged {
				result = arrangementCount(p: p + gs + 1, g: g + 1)
			}
			else {
				result = 0 // not a solution - we must always start damaged group here
			}
		}
		else if conditionRecord[p] == .operational {
			result = arrangementCount(p: p+1, g: g) // operational spring -> go to the next spring
		}

		_memo[mRec] = result
		return result
	}
	
	func expanded(times: Int) -> SpringRecord {
		var condition = [String]()
		var groups = [String]()
		for _ in 0..<times {
			condition.append(originalConditionRecord)
			groups.append(originalGroupLengths)
		}
		let defn = "\(condition.joined(separator: "?")) \(groups.joined(separator: ","))"
		return SpringRecord(s: defn)
	}
    
    var debugDescription: String {
		return "\(conditionRecord.map({$0.stringRepresentation}).joined()) \(groupLengths)"
    }
}

File: Perl/day12.pl
#!/usr/bin/env perl
use v5.42;
# use feature 'class';
# no warnings qw( experimental::class );

our $directory;
BEGIN { use Cwd; $directory = cwd; }
use lib $directory . '/lib';

use feature 'signatures';
use Data::Printer;
use List::Util qw(any sum);
use Math::Combinatorics;

use AOC::Util;

my $DAY = sprintf("%02d", 12);
my $INPUT = $ARGV[0];
my $INPUT_FILE = 'day' . $DAY . '_' . $INPUT . '.txt';

my @input = read_input("../Input/$INPUT_FILE");

say "Advent of Code 2023, Day 12: Hot Springs";

my @records = parse_records(1, @input);
# p @records;
solve_part_one(@records);

@records = parse_records(5, @input);
solve_part_two(@records);

exit( 0 );

sub solve_part_one(@records) {
	my @counts = map { get_possible_arrangement_count( $_ ) } @records;
	my $arr_count = sum(@counts);
	say "Part One: the total number of possible arrangements is $arr_count.";
}

sub solve_part_two(@input) {
	my @counts = map { get_possible_arrangement_count( $_ ) } @records;
	my $arr_count = sum(@counts);
	say "Part Two: the total number of possible unfolded arrangements is $arr_count.";
}

my %memo = ();
sub get_possible_arrangement_count($record) { # hash ref
	%memo = ();
	my $count = get_arr_count($record, 0, 0);
	return $count;
}

sub get_arr_count($record, $p, $g) {
	my @groups = @{$record->{'groups'}};
	my @status = @{$record->{'status'}};
	if ($g >= scalar @groups) { # no more groups
		if ($p < $#status and (any {$_ eq '#'} @status[$p..$#status])) {
			return 0; # not a solution - there are still damaged springs in the record
		}
		return 1;
	}

	my $g_length = $groups[$g];
	if ($p + $g_length > scalar @status) {
		return 0; # we ran out of springs but there are still groups to arrange
	}

	my $memo_key = "$p,$g";
	if (exists $memo{$memo_key}) { return $memo{$memo_key}; }

	my $result = 0;

	if ($status[$p] eq '?') {
		# if we can start group of damaged springs here
		# eg: '??#...... 3' we can place 3 '#' and there is '?' or '.' after the group
		# eg: '??##...... 3' we cannot place 3 '#' here
		my $b_no_operational = !any {$_ eq '.'} @status[$p..$p+$g_length-1];
		no warnings 'uninitialized';
		my $b_next_not_broken = $status[$p + $g_length] ne "#";
		use warnings 'uninitialized';
		if ($b_no_operational and $b_next_not_broken) {
			# start damaged group here + this spring is operational ('.')
			$result = get_arr_count($record, $p + $g_length + 1, $g + 1) +
				get_arr_count($record, $p + 1, $g);
		}
		else {
			# This spring is operational
			$result = get_arr_count($record, $p+1, $g);
		}
	}
	elsif ($status[$p] eq '#') {
		# if we can start damaged group here
		my $b_no_operational = !any {$_ eq '.'} @status[$p..$p+$g_length-1];
		no warnings 'uninitialized';
		my $b_next_not_broken = ($status[$p + $g_length] ne "#");
		use warnings 'uninitialized';
		if ($b_no_operational and $b_next_not_broken) {
			$result = get_arr_count($record, $p + $g_length + 1, $g + 1);
		}
		else {
			$result = 0; # not a solution - we must always start damaged group here
		}
	}
	elsif ($status[$p] eq '.') {
		# operational spring -> go to the next spring
		$result = get_arr_count($record, $p + 1, $g);
	}

	$memo{$memo_key} = $result;
	return $result;
}

sub parse_records($folds, @input) {
	my @recs = ();
	for my $line (@input) {
		my @parts = split(/ /, $line);
		my $status_str = join('?', ($parts[0]) x $folds);
		my @status = split(//, $status_str);
		push(@status, '.');
		my $group_str = join(',', ($parts[1]) x $folds);
		my @groups = split(/,/, $group_str);
		my %info = ('status' => \@status, 'groups' => \@groups);
		push(@recs, \%info);
	}
	return @recs;
}
